### PR TITLE
Round drop_first to be divisible by num_physical_nodes.

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:

This is to handle a resumption edge case: the user trains on N nodes, checkpoints and resumes training on N / K nodes, then checkpoints and resumes at an unfortunate timestep back on N nodes which is not divisible by N.  We round down drop_first to the nearest divisible value, potentially re-including a very small selection of already seen samples.

## Issue #, if available:
STR-83

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: tests/test_partition.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    num_canonical_nodes = 176
+    num_physical_nodes = 22
+    ranks_per_node = 8
+    workers_per_rank = 8
+    batch_size = 10
+    drop_first = 408112000
+    """
+
+    for drop_first in range(0, 500):
+        get_partitions(partition_algo, num_samples, num_canonical_nodes, num_physical_nodes,
```
Comment (karan6181) [inline tests/test_partition.py:29]: What are we validating here ? Can we do some assertion here w.r.t shape, total samples coming out of `get_partitions()` ?

File: streaming/base/partition/orig.py [hunk 2]
```diff
@@ -52,6 +52,9 @@ def get_partitions_orig(num_samples: int,
 
     batch_size = batch_size or 1
 
+    # If drop_first isn't a multiple of num_physical_nodes, round down to make it divisible.
+    drop_first -= drop_first % num_physical_nodes
```
Comment (karan6181) [inline streaming/base/partition/orig.py:56]: If the drop_first isn't a multiple of num_physical_nodes, does this mean that there would a certain group of sample(s) that would be repeated (before checkpoint and after checkpoint) ?  If yes, then can we add a logger.warning message saying that some sample(s) will be repeated ?

File: tests/test_partition.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    num_samples = 1999990222
+    num_canonical_nodes = 176
+    num_physical_nodes = 22
+    ranks_per_node = 8
+    workers_per_rank = 8
+    batch_size = 10
+    drop_first = 408112000
+    """
+
+    for drop_first in range(0, 500):
```
[the remark this answers is not in the document]
Comment (knighton) [reply tests/test_partition.py:28]: it's instant, as num_samples is small

[the remark this answers is not in the document]
Comment (karan6181) [reply tests/test_partition.py:28]: Ok. Can you share the output of `pytest -vv -s tests/test_partition.py` ? Thanks!

[the remark this answers is not in the document]
Comment (knighton) [reply tests/test_partition.py:28]: ============================================= 1 passed, 1 warning in 0.45s =============================================

File: tests/test_partition.py (hunk 1 above)
Comment (knighton) [reply tests/test_partition.py]: Added shape check.

File: tests/test_partition.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        x = get_partitions(partition_algo, num_samples, num_canonical_nodes, num_physical_nodes,
+                           ranks_per_node, workers_per_rank, batch_size, drop_first)
+        assert x.shape == (22, 8, 8, 1, 10)
```
Comment (karan6181) [inline tests/test_partition.py:31]: Thanks! One more request, can you also validate the number of samples (after drop_first) to validate that we are not dropping unnecessary samples ? I think based on the above shape `(22, 8, 8, 1, 10)`, most of the samples are the placeholder (-1). So validate the actual number of samples after drop_first would be good testing too!

File: streaming/base/partition/orig.py (hunk 2 above)
Comment (karan6181) [reply streaming/base/partition/orig.py]: Addressed as part of https://github.com/mosaicml/streaming/pull/301/commits/86b150a00025a92f1db686b0c85732c3f50591a5.

File: streaming/base/partition/orig.py [hunk 5]
```diff
@@ -52,6 +59,12 @@ def get_partitions_orig(num_samples: int,
 
     batch_size = batch_size or 1
 
+    # If drop_first isn't a multiple of num_physical_nodes, round down to make it divisible.
+    if drop_first % num_physical_nodes:
+        logger.warning('`drop_first` was not divisible by `num_physical_nodes`. Rounding it ' +
```
Comment (karan6181) [inline streaming/base/partition/orig.py:64]: Can you change `logger.warning` to `warnings.warn` and remove the changes that you made in pytest.ini file ?

Comment (knighton) [reply streaming/base/partition/orig.py]: done

**Review by karan6181 — APPROVED** [review]

**Pull request merged**